modules/tensor_mechanics/src/materials/ComputeFiniteBeamStrain.C

```diff
@@ -0,0 +1,123 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "ComputeFiniteBeamStrain.h"
+#include "Assembly.h"
+#include "NonlinearSystem.h"
+#include "MooseVariable.h"
+
+#include "libmesh/quadrature.h"
+#include "libmesh/utility.h"
+
+registerMooseObject("TensorMechanicsApp", ComputeFiniteBeamStrain);
+
+template <>
+InputParameters
+validParams<ComputeFiniteBeamStrain>()
+{
+  InputParameters params = validParams<ComputeIncrementalBeamStrain>();
+  params.addClassDescription("Compute a rotation increment for finite rotations of the beam and "
+                             "computes the small/large strain increments in the current rotated "
+                             "configuration of the beam.");
+  return params;
+}
+
+ComputeFiniteBeamStrain::ComputeFiniteBeamStrain(const InputParameters & parameters)
+  : ComputeIncrementalBeamStrain(parameters),
+    _total_rotation_old(getMaterialPropertyOld<RankTwoTensor>("total_rotation"))
+{
+}
+
+void
+ComputeFiniteBeamStrain::computeProperties()
+{
+  // Calculation of rotation matrix from original beam local configuration to configuration at time
+  // t
+  computeRotation();
+
+  ComputeIncrementalBeamStrain::computeProperties();
+}
+
+void
+ComputeFiniteBeamStrain::computeRotation()
+{
+  // First - calculate difference in beam displacement between node 1 and node 0 in global
+  // coordinate system
+
+  // fetch the two end nodes for _current_elem
+  std::vector<Node *> node;
+  for (unsigned int i = 0; i < 2; ++i)
+    node.push_back(_current_elem->get_node(i));
+
+  // compute original length of beam element
+  RealGradient dxyz;
+  for (unsigned int i = 0; i < _ndisp; ++i)
+    dxyz(i) = (*node[1])(i) - (*node[0])(i);
+
+  Real original_length = dxyz.norm();
+
+  // Fetch the solution for the two end nodes at the current and previous time step
+  NonlinearSystemBase & nonlinear_sys = _fe_problem.getNonlinearSystemBase();
+  const NumericVector<Number> & sol = *nonlinear_sys.currentSolution();
+  const NumericVector<Number> & sol_old = nonlinear_sys.solutionOld();
+
+  RealVectorValue disp0(3, 0.0), disp1(3, 0.0);
+  RealVectorValue rot0(3, 0.0), rot1(3, 0.0);
+  RealVectorValue delta_disp(3, 0.0);
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
```

Comment:
Since you're reusing these indices so many times here, probably better to create a local variable that is set once and re-used many times.